Tests/GRPCTests/MutualTLSTests.swift

```diff
@@ -0,0 +1,269 @@
+  var server: Server?
+
+  override func setUp() {
+    super.setUp()
+    self.serverEventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+    self.clientEventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+  }
+
+  override func tearDown() {
+    XCTAssertNoThrow(try self.channel?.close().wait())
+    XCTAssertNoThrow(try self.server?.close().wait())
+    XCTAssertNoThrow(try self.clientEventLoopGroup.syncShutdownGracefully())
+    XCTAssertNoThrow(try self.serverEventLoopGroup.syncShutdownGracefully())
+    super.tearDown()
+  }
+
+  func performTestWith(
+    _ serverTLSConfiguration: GRPCTLSConfiguration?,
+    _ clientTLSConfiguration: GRPCTLSConfiguration?,
+    expectServerHandshakeError: Bool,
+    expectedClientError: ExpectedClientError?
+  ) throws {
+    // Setup the server.
+    var serverConfiguration = Server.Configuration.default(
+      target: .hostAndPort("localhost", 0),
+      eventLoopGroup: self.serverEventLoopGroup,
+      serviceProviders: [EchoProvider()]
+    )
+    serverConfiguration.tlsConfiguration = serverTLSConfiguration
+    serverConfiguration.logger = self.serverLogger
+    let serverErrorExpectation = self.expectation(description: "server error")
+    serverErrorExpectation.isInverted = !expectServerHandshakeError
+    serverErrorExpectation.assertForOverFulfill = false
+    let serverErrorDelegate = ServerErrorRecordingDelegate(expectation: serverErrorExpectation)
+    serverConfiguration.errorDelegate = serverErrorDelegate
+
+    self.server = try! Server.start(configuration: serverConfiguration).wait()
+
+    let port = self.server!.channel.localAddress!.port!
+
+    // Setup the client.
+    var clientConfiguration = ClientConnection.Configuration.default(
+      target: .hostAndPort("localhost", port),
+      eventLoopGroup: self.clientEventLoopGroup
+    )
+    clientConfiguration.tlsConfiguration = clientTLSConfiguration
+    clientConfiguration.connectionBackoff = nil
+    clientConfiguration.backgroundActivityLogger = self.clientLogger
+    let clientErrorExpectation = self.expectation(description: "client error")
+    switch expectedClientError {
+    case .none:
+      clientErrorExpectation.isInverted = true
+    case .handshakeError, .alertCertRequired:
+      // After the SSL error, the connection being closed also presents as an error.
+      clientErrorExpectation.expectedFulfillmentCount = 2
+    case .dropped:
+      clientErrorExpectation.expectedFulfillmentCount = 1
+    }
+    let clientErrorDelegate = ErrorRecordingDelegate(expectation: clientErrorExpectation)
+    clientConfiguration.errorDelegate = clientErrorDelegate
+
+    self.channel = ClientConnection(configuration: clientConfiguration)
+    let client = Echo_EchoClient(channel: channel!)
+
+    // Make the call.
+    let call = client.get(.with { $0.text = "mumble" })
+
+    // Wait for side effects.
+    self.wait(for: [clientErrorExpectation, serverErrorExpectation], timeout: 1)
```

Comment:
CI can be pretty slow at times, so best to have a fairly large timeout:

```suggestion
    self.wait(for: [clientErrorExpectation, serverErrorExpectation], timeout: 10)
```